src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAwaitForeachTests.cs

```diff
@@ -1695,12 +1695,26 @@ public System.Threading.Tasks.Task<bool> MoveNextAsync()
             => throw null;
         public int Current { get => throw null; }
     }
-}";
-            var comp = CreateCompilationWithTasksExtensions(source + s_IAsyncEnumerable);
+}" + s_IAsyncEnumerable;
+
+            var comp = CreateCompilationWithTasksExtensions(source, parseOptions: TestOptions.Regular12);
             comp.VerifyDiagnostics(
-                // (6,32): error CS8177: Async methods cannot have by-reference locals
+                // (6,32): error CS8652: The feature 'Ref and unsafe in async and iterator methods' is currently in Preview and *unsupported*. To use Preview features, use the 'preview' language version.
                 //         await foreach (ref var i in new C())
-                Diagnostic(ErrorCode.ERR_BadAsyncLocalType, "i").WithLocation(6, 32));
+                Diagnostic(ErrorCode.ERR_FeatureInPreview, "i").WithArguments("Ref and unsafe in async and iterator methods").WithLocation(6, 32));
+
+            var expectedDiagnostics = new[]
+            {
+                // (6,37): error CS1510: A ref or out value must be an assignable variable
+                //         await foreach (ref var i in new C())
+                Diagnostic(ErrorCode.ERR_RefLvalueExpected, "new C()").WithLocation(6, 37)
+            };
+
+            comp = CreateCompilationWithTasksExtensions(source, parseOptions: TestOptions.RegularNext);
+            comp.VerifyDiagnostics(expectedDiagnostics);
+
+            comp = CreateCompilationWithTasksExtensions(source);
+            comp.VerifyDiagnostics(expectedDiagnostics);
```

Comment:
>comp [](http://example.com/codeflow?start=12&length=4)

Are we testing a `ref` local with async foreach and iterators?
```csharp
using System.Collections.Generic;
using System.Threading.Tasks;

class C
{
    static async void Main()
    {
        await foreach (int i in F())
        {
            Console.WriteLine(i);
        }
    }

    static async IAsyncEnumerable<int> F()
    {
        await foreach (ref var i in new C())
        {
            yield return i;
        }
    }
    
    public Enumerator GetAsyncEnumerator(System.Threading.CancellationToken token = default) => new();
    
    public sealed class Enumerator
    {
        private readonly int[] _array = [1, 2, 3];
        private int _index;
        public System.Threading.Tasks.Task<bool> MoveNextAsync()
        {
            if (_index < _array.Length) _index++;
            return Task.FromResult(_index < _array.Length);
        }       
        public ref int Current => ref _array[_index];
    }
}
```

Reply:
Probably not, will add, thanks.